proof/refine/ARM/ArchKHeap_R.thy

```diff
@@ -0,0 +1,438 @@
+                            swp_def fun_upd_def obj_at_def)
+  apply (subst conj_assoc[symmetric])
+  apply (extract_conjunct \<open>match conclusion in "ghost_relation _ _ _" \<Rightarrow> -\<close>)
+   apply (clarsimp simp add: ghost_relation_def)
+   apply (erule_tac x=ptr in allE)+
+   apply (clarsimp simp: obj_at_def a_type_def
+                   split: Structures_A.kernel_object.splits if_split_asm)
+   apply (simp split: arch_kernel_obj.splits if_splits)
+  apply (fold fun_upd_def)
+  apply (simp only: pspace_relation_def pspace_dom_update dom_fun_upd2 simp_thms)
+  apply (elim conjE)
+  apply (frule bspec, erule domI)
+  apply (prop_tac "is_ArchObj ko", clarsimp simp: a dest!: a_type_eq_is_ArchObj)
+  apply (prop_tac "typ_at' (koTypeOf (injectKO ob')) ptr b")
+   subgoal
+     by (clarsimp simp: typ_at'_def ko_wp_at'_def obj_at'_def projectKO_opts_defs
+                        is_other_obj_relation_type_def a_type_def other_aobj_relation_def
+                 split: Structures_A.kernel_object.split_asm if_split_asm
+                        arch_kernel_obj.split_asm kernel_object.split_asm
+                        arch_kernel_object.split_asm)
+  apply clarsimp
+  apply (rule conjI)
+   apply (rule ballI, drule(1) bspec)
+   apply (drule domD)
+   apply (clarsimp simp: is_other_obj_relation_type t a)
+   apply (drule(1) bspec)
+   apply clarsimp
+   apply (frule_tac ko'=ko and x'=ptr in obj_relation_cut_same_type)
+   apply ((fastforce simp add: is_other_obj_relation_type t)+)[3] (* loops when folded into above *)
+   apply (insert t)
+   apply ((erule disjE
+          | clarsimp simp: is_other_obj_relation_type is_other_obj_relation_type_def a_type_def)+)[1]
+  \<comment> \<open>ready_queues_relation\<close>
+  apply (prop_tac "koTypeOf (injectKO ob') \<noteq> TCBT")
+   subgoal
+     by (clarsimp simp: other_aobj_relation_def; cases ob; cases "injectKO ob'";
+         simp split: arch_kernel_obj.split_asm)
+  by (fastforce dest: tcbs_of'_non_tcb_update)
+
+lemmas [KHeap_R_assms] =
+  setObject_other_corres[where 'a=endpoint]
+  setObject_other_corres[where 'a=notification]
+
+lemma dmo_storeWordVM' [simp]:
+  "doMachineOp (storeWordVM x y) = return ()"
+  by (simp add: storeWordVM_def)
+
+(* only on arches without kernel mappings *)
+lemma pspace_in_kernel_mappings'_wp[wp]:
+  "\<lbrace>\<top>\<rbrace> f \<lbrace>\<lambda>_. pspace_in_kernel_mappings'\<rbrace>"
+  unfolding pspace_in_kernel_mappings'_def
+  by wp
+
+(* FIXME arch-split: check with other arches what is actually in common, standardise *)
+lemma setEndpoint_pspace_in_kernel_mappings'[KHeap_R_assms]:
+  "setEndpoint p ko \<lbrace>pspace_in_kernel_mappings'\<rbrace>"
+  by wp
+
+lemma setNotification_pspace_in_kernel_mappings'[KHeap_R_assms]:
+  "setNotification p ko \<lbrace>pspace_in_kernel_mappings'\<rbrace>"
+  by wp
+
+(* FIXME arch-split: possible interface lemma, but can't be done via locale *)
+lemma valid_global_refs_lift':
```

Comment:
(convert from FIXME to normal comment)

Reply:
as in `interface lemma, but can't be done via locale`?

Reply:
done